### PR TITLE
Add rule SARIF2004.OptimizeFileSize.EliminateIdOnlyRules

### PR DESCRIPTION
**Pull request by harleenkohli**

### CODE REVIEW

File: src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs [hunk 1]
```diff
@@ -23,12 +24,19 @@ public class OptimizeFileSize : SarifValidationSkimmerBase
         public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.SARIF2004_OptimizeFileSize_FullDescription_Text };
 
         protected override IEnumerable<string> MessageResourceNames => new string[] {
-                    nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateLocationOnlyArtifacts_Text)
+                    nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateLocationOnlyArtifacts_Text),
+                                        nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateIdOnlyRules_Text)
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs:28]: >                     [](start = 20, length = 20)

Alignment. Tabs/spaces issue?

File: src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs [hunk 2]
```diff
@@ -69,10 +77,46 @@ private bool HasResultLocationsWithUriAndIndex(string resultPointer)
 
         private bool HasLocationOnlyArtifacts(string artifactPointer)
         {
-            JToken artifactToken = artifactPointer.ToJToken(Context.InputLogToken);
+            var artifactToken = artifactPointer.ToJToken(Context.InputLogToken);
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs:80]: >var [](start = 12, length = 3)

Revert. Even though the name `ToJToken` is quite clear, still our standard is only to use `var` when the right-hand side is `new T`, `(T)expr`, or `expr as T`; the type must be absolutely explicit.

File: src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs [hunk 3]
```diff
@@ -69,10 +77,46 @@ private bool HasResultLocationsWithUriAndIndex(string resultPointer)
...
             return
                 artifactToken.HasProperty(SarifPropertyName.Location) &&
                 artifactToken.Children().Count() == 1;
         }
+
+        private void AnalyzeIdOnlyRules(Run run, string runPointer)
+        {
+            // We only verify first item in the rules array,
+            // since tools will typically generate similar nodes.
+            // This approach may cause occasional false negatives.
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs:90]: >// [](start = 12, length = 2)

Remove extra blank lines below.

File: src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs [hunk 4]
```diff
@@ -69,10 +77,46 @@ private bool HasResultLocationsWithUriAndIndex(string resultPointer)
...
+
+            ReportingDescriptor firstRule = run?.Tool?.Driver?.Rules?.FirstOrDefault();
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs:92]: >run?.Tool?.Driver? [](start = 44, length = 18)

You don't need to test `run` because the visitor won't call the `Analyze` method with a null run. And `Tool` and `Driver` are required, so the "schema validation" phase of the validator tool will catch this before we ever get here. So `run.Tool.Driver?.Rules?.FirstOrDefault()` will do.

File: src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs [hunk 5]
```diff
@@ -69,10 +77,46 @@ private bool HasResultLocationsWithUriAndIndex(string resultPointer)
...
+                    nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateIdOnlyRules_Text));
+            }
+        }
+
+        private bool HasIdOnlyRules(string rulePointer)
+        {
+            var ruleToken = rulePointer.ToJToken(Context.InputLogToken);
+            return
+                ruleToken.HasProperty(SarifPropertyName.Id) &&
+                ruleToken.Children().Count() == 1;
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs:119]: >1 [](start = 48, length = 1)

Good.

File: src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs (hunk 1 above)
Comment (harleenkohli) [reply src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs:28]: eww.. thanks. yes 

---
In reply to: [447981759](https://github.com/microsoft/sarif-sdk/pull/1950#discussion_r447981759) [](ancestors = 447981759)

File: src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs (hunk 2 above)
Comment (harleenkohli) [reply src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs:80]: Okay :) 

---
In reply to: [447982484](https://github.com/microsoft/sarif-sdk/pull/1950#discussion_r447982484) [](ancestors = 447982484)

File: src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs (hunk 4 above)
Comment (harleenkohli) [reply src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs:92]: oh correct! right and right, they mandatory!

---
In reply to: [447983469](https://github.com/microsoft/sarif-sdk/pull/1950#discussion_r447983469) [](ancestors = 447983469)

Comment (harleenkohli) [reply src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs]: note: it should be run.Tool.Driver.Rules?.FirstOrDefault() 

---
In reply to: [447987160](https://github.com/microsoft/sarif-sdk/pull/1950#discussion_r447987160) [](ancestors = 447987160,447983469)
